# Common File

In [29]:
LIBRARY='rdkit'

In [30]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib as mpl
from tqdm import tqdm
import pickle
from IPython.display import clear_output

In [31]:
# np.random.seed(200) #-- works for lin reg
# np.random.seed(110)

# Data

In [32]:
data_train=pd.read_csv('DATA_train.csv')
data_test=pd.read_csv('DATA_test.csv')
data_train.shape,data_test.shape

((129, 3857), (56, 3857))

In [33]:
data_train

Sr. No. FG_name FG Position  Redox Potential mol_filename_neutral  \
0       148      OH       1 2 8            -1.82      OH_R128_neutral   
1        13      SH           1            -1.66        SH_R1_neutral   
2       152      OH       2 3 7            -1.93      OH_R237_neutral   
3        34    COOH           1            -1.30      COOH_R1_neutral   
4        24      Cl           2            -1.63        Cl_R2_neutral   
..      ...     ...         ...              ...                  ...   
124      45      CN         1 6            -1.14       CN_R16_neutral   
125      40     NO2           1            -1.35       NO2_R1_neutral   
126     112     NH2       1 4 7            -2.05     NH2_R147_neutral   
127     183     NO2     1 2 4 8            -0.30    NO2_R1248_neutral   
128     178     NO2       1 4 7            -0.70     NO2_R147_neutral   

     FG_no_2d_basic  FG_position_1_2d_basic  FG_position_2_2d_basic  \
0                17                     1.0                     1.0   
1                20                     1.0                     0.0   
2                17                     0.0                     1.0   
3                11                     1.0                     0.0   
4                12                     0.0                     1.0   
..              ...                     ...                     ...   
124               7                     1.0                     0.0   
125              15                     1.0                     0.0   
126              14                     1.0                     0.0   
127              15                     1.0                     1.0   
128              15                     1.0                     0.0   

     FG_position_3_2d_basic  FG_position_4_2d_basic  ...  rdkit_fp_2035  \
0                       0.0                     0.0  ...              0   
1                       0.0                     0.0  ...              0   
2                       1.0                     0.0  ...              0   
3                       0.0                     0.0  ...              0   
4                       0.0                     0.0  ...              0   
..                      ...                     ...  ...            ...   
124                     0.0                     0.0  ...              0   
125                     0.0                     0.0  ...              0   
126                     0.0                     1.0  ...              0   
127                     0.0                     1.0  ...              1   
128                     0.0                     1.0  ...              1   

     rdkit_fp_2036  rdkit_fp_2037  rdkit_fp_2038  rdkit_fp_2039  \
0                0              0              0              0   
1                0              0              0              0   
2                0              0              0              0   
3                0              0              0              0   
4                0              0              0              1   
..             ...            ...            ...            ...   
124              0              1              0              0   
125              0              1              0              0   
126              0              1              0              0   
127              0              1              1              0   
128              0              1              1              0   

     rdkit_fp_2042  rdkit_fp_2043  rdkit_fp_2044  rdkit_fp_2046  rdkit_fp_2047  
0                0              0              0              0              0  
1                1              0              0              0              0  
2                0              0              0              0              0  
3                0              0              0              0              0  
4                0              0              0              0              0  
..             ...            ...            ...            ...            ...  
1

In [34]:
data_test

Sr. No.  FG_name  FG Position  Redox Potential mol_filename_neutral  \
0        65       CN        2 3 7            -0.89      CN_R237_neutral   
1        12  S(CH3)2            2            -1.77   S(CH3)2_R2_neutral   
2        96      NH2          1 8            -2.02      NH2_R18_neutral   
3       184      NO2      1 2 4 9            -0.34    NO2_R1249_neutral   
4        10  P(CH3)2            2            -1.75   P(CH3)2_R2_neutral   
5       155       OH      1 2 3 7            -1.87     OH_R1237_neutral   
6       153       OH      1 2 3 4            -1.78     OH_R1234_neutral   
7        17     C6H5            1            -1.76      C6H5_R1_neutral   
8       172      NO2          1 9            -1.14      NO2_R19_neutral   
9         4      NH2            2            -1.92       NH2_R2_neutral   
10      129      NH2      1 3 6 8            -2.25    NH2_R1368_neutral   
11      136      NH2  1 2 3 7 8 9            -2.36  NH2_R123789_neutral   
12       77       CN      1 2 7 8            -0.58     CN_R1278_neutral   
13       43       CN          1 3            -1.10       CN_R13_neutral   
14      163       OH      1 4 7 8            -1.55     OH_R1478_neutral   
15       32   COOCH3            1            -1.62    COOCH3_R1_neutral   
16      138      NH2  1 2 3 4 6 9            -2.21  NH2_R123469_neutral   
17       15   CH2=CH            1            -1.73    CH2=CH_R1_neutral   
18       14       SH            2            -1.72        SH_R2_neutral   
19       99      NH2          2 7            -2.05      NH2_R27_neutral   
20      188      NO2  1 2 3 4 6 9             0.16  NO2_R123469_neutral   
21       89       CN  2 3 6 7 8 9            -0.06   CN_R236789_neutral   
22       75       CN      1 2 6 7            -0.52     CN_R1267_neutral   
23      107      NH2        1 3 6            -2.13     NH2_R136_neutral   
24       47       CN          1 8            -1.14       CN_R18_neutral   
25       18     C6H5            2            -1.73      C6H5_R2_neutral   
26      150       OH        1 3 8            -1.82      OH_R138_neutral   
27      175      NO2        1 2 3            -0.74     NO2_R123_neutral   
28      164       OH      2 3 7 8            -1.99     OH_R2378_neutral   
29       23       Cl            1            -1.61        Cl_R1_neutral   
30      105      NH2        1 2 8            -2.09     NH2_R128_neutral   
31      165       OH  1 2 3 6 7 8            -1.90   OH_R123678_neutral   
32       86       CN  1 2 3 4 6 9             0.02   CN_R123469_neutral   
33      133      NH2      1 4 7 8            -2.27    NH2_R1478_neutral   
34      176      NO2        1 2 4            -0.56     NO2_R124_neutral   
35       94      NH2          1 6            -1.96      NH2_R16_neutral   
36       39       CN            2            -1.43        CN_R2_neutral   
37       56       CN        1 2 8            -0.80      CN_R128_neutral   
38        7     OCH3            1            -1.75      OCH3_R1_neutral   
39      125      NH2      1 2 6 9            -2.07    NH2_R1269_neutral   
40       33   COOCH3            2            -1.57    COOCH3_R2_neutral   
41      101      NH2        1 2 3            -2.04     NH2_R123_neutral   
42      140       OH          1 2            -1.84       OH_R12_neutral   
43       72       CN      1 2 4 7            -0.45     CN_R1247_neutral   
44       80       CN      1 3 6 8            -0.57     CN_R1368_neutral   
45       93      NH2          1 4            -1.89      NH2_R14_neutral   
46       28    COCH3            1            -1.58     COCH3_R1_neutral   
47       48       CN          1 9            -1.15       CN_R19_neutral   
48      130      NH2      1 3 6 9            -2.34    NH2_R1369_neutral   
49      144       OH          2 3            -1.93       OH_R23_neutral   
50       44       CN          1 4            -1.06       CN_R14_neutral   
51      106      NH2        1 2 9            -2.02     NH2_R129_neutral   
52      179      NO2        2

# Preprocessing

In [35]:
from sklearn.model_selection import train_test_split

X_train=data_train.drop(columns=['Sr. No.','FG_name','FG Position','Redox Potential','mol_filename_neutral'])
X_test=data_test.drop(columns=['Sr. No.','FG_name','FG Position','Redox Potential','mol_filename_neutral'])
y_train=data_train['Redox Potential']
y_test=data_test['Redox Potential']

print(X_train.shape,y_train.shape)
print(X_test.shape,y_test.shape)

(129, 3852) (129,)
(56, 3852) (56,)


In [36]:
from sklearn.preprocessing import StandardScaler

numerical_cols=[col for col in X_train.columns if '_2d_rdkit' in col]
numerical_cols+=[col for col in X_train.columns if '_3d_rdkit' in col]
numerical_cols+=[col for col in X_train.columns if '_2d_mordred' in col]
numerical_cols+=[col for col in X_train.columns if '_3d_mordred' in col]
X_train_scaled=X_train.copy()
X_test_scaled=X_test.copy()
for col in numerical_cols:
    scaler=StandardScaler().fit(X_train[col].values.reshape(-1,1))
    X_train_scaled[col]=scaler.transform(X_train[col].values.reshape(-1,1))
    X_test_scaled[col]=scaler.transform(X_test[col].values.reshape(-1,1))

In [66]:
feat_cols_2d_3d_fp=[col for col in X_train_scaled.columns if f'_2d_{LIBRARY}' in col]
feat_cols_2d_3d_fp+=[col for col in X_train_scaled.columns if f'_3d_{LIBRARY}' in col]
feat_cols_2d_3d_fp+=[col for col in X_train_scaled.columns if '_fp_' in col]
X_train_scaled_2d_3d_fp=X_train_scaled[feat_cols_2d_3d_fp]
X_test_scaled_2d_3d_fp=X_test_scaled[feat_cols_2d_3d_fp]

feat_cols_2d_3d=[col for col in X_train_scaled.columns if f'_2d_{LIBRARY}' in col]
feat_cols_2d_3d+=[col for col in X_train_scaled.columns if f'_3d_{LIBRARY}' in col]
X_train_scaled_2d_3d=X_train_scaled[feat_cols_2d_3d]
X_test_scaled_2d_3d=X_test_scaled[feat_cols_2d_3d]

feat_cols_2d=[col for col in X_train_scaled.columns if f'_2d_{LIBRARY}' in col]
X_train_scaled_2d=X_train_scaled[feat_cols_2d]
X_test_scaled_2d=X_test_scaled[feat_cols_2d]

feat_cols_3d=[col for col in X_train_scaled.columns if f'_3d_{LIBRARY}' in col]
X_train_scaled_3d=X_train_scaled[feat_cols_3d]
X_test_scaled_3d=X_test_scaled[feat_cols_3d]

feat_cols_fp=[col for col in X_train_scaled.columns if f'_fp_' in col]
X_train_scaled_fp=X_train_scaled[feat_cols_fp]
X_test_scaled_fp=X_test_scaled[feat_cols_fp]

# Examples

## - Baseline Models

### 1. Linear Regression

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

reg = LinearRegression().fit(X_train_scaled, y_train)

y_pred=reg.predict(X_test_scaled)
r2=r2_score(y_test,y_pred)

y_pred_train=reg.predict(X_train_scaled)
r2_train=r2_score(y_train,y_pred_train)

print(f'R^2 = {r2}')
plt.scatter(y_test,y_pred,label=f'Test R^2 = {r2}')
plt.plot(np.linspace(-2,0),np.linspace(-2,0),ls='dashed',label='X=Y line')
plt.xlabel('DFT Potential')
plt.ylabel('ML Prediction')
plt.title('Linear Regression (Test-set)')

plt.scatter(y_train,y_pred_train,label=f'Train R^2 = {r2_train}')

plt.legend()

### 2. SVR

In [ ]:
from sklearn.svm import SVR

svr=SVR(C=2, epsilon=0.2).fit(X_train_scaled, y_train)

y_pred=svr.predict(X_test_scaled)
r2=r2_score(y_test,y_pred)


y_pred_train=svr.predict(X_train_scaled)
r2_train=r2_score(y_train,y_pred_train)

print(f'R^2 = {r2}')
plt.scatter(y_test,y_pred,label=f'Test R^2 = {r2}')
plt.plot(np.linspace(-2,0),np.linspace(-2,0),ls='dashed',label='X=Y line')
plt.xlabel('DFT Potential')
plt.ylabel('ML Prediction')
plt.title('SVR (Test-set)')

plt.scatter(y_train,y_pred_train,label=f'Train R^2 = {r2_train}')

plt.legend()

### 3. Gradient Boosting Regression

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor

gbr=GradientBoostingRegressor().fit(X_train_scaled, y_train)

y_pred=gbr.predict(X_test_scaled)
r2=r2_score(y_test,y_pred)

y_pred_train=gbr.predict(X_train_scaled)
r2_train=r2_score(y_train,y_pred_train)

print(f'R^2 = {r2}')
plt.scatter(y_test,y_pred,label=f'Test R^2 = {r2}')
plt.plot(np.linspace(-2,0),np.linspace(-2,0),ls='dashed',label='X=Y line')
plt.xlabel('DFT Potential')
plt.ylabel('ML Prediction')
plt.title('Gradient Boosting Regression (Test-set)')

plt.scatter(y_train,y_pred_train,label=f'Train R^2 = {r2_train}')

plt.legend()

## - Hyper-Parameter Tuning (Kernel Ridge Regression)

In [ ]:
from IPython import display
from sklearn.model_selection import GridSearchCV
from sklearn.kernel_ridge import KernelRidge

param_grid={'kernel': ['chi2', 'linear','polynomial', 'rbf', 'laplacian', 'sigmoid', 'cosine'],
            'gamma': [1e-8,1e-7,1e-6,1e-5,1e-3, 1e-4],
            'alpha':np.linspace(1e-4,1,10)}

scoring = ['neg_mean_squared_error','r2']

estimator = KernelRidge()

krr_gs = GridSearchCV(estimator,
                  param_grid=param_grid,
                  scoring=scoring,
                  refit='neg_mean_squared_error',
                  return_train_score=True,
                  cv=10,
                  verbose=1,
                  n_jobs=-1
                 )
krr_gs.fit(X_train_scaled, y_train)
display.clear_output()
print(krr_gs.best_params_)

plt.figure()
krr_reg = KernelRidge(kernel='rbf').fit(X_train_scaled, y_train)
y_pred=krr_reg.predict(X_test_scaled)
r2=r2_score(y_test,y_pred)
print(f'R^2 = {r2}')
plt.scatter(y_test,y_pred,label=f'R^2 = {r2}')
plt.plot(np.linspace(-2,0),np.linspace(-2,0),ls='dashed',label='X=Y line')
plt.xlabel('DFT Potential')
plt.ylabel('ML Prediction')
plt.title('Unoptimized Kernel Ridge Regression (Test-set)')
plt.legend()

plt.figure()
y_pred=krr_gs.best_estimator_.predict(X_test_scaled)
r2=r2_score(y_test,y_pred)
print(f'R^2 = {r2}')
plt.scatter(y_test,y_pred,label=f'R^2 = {r2}')
plt.plot(np.linspace(-2,0),np.linspace(-2,0),ls='dashed',label='X=Y line')
plt.xlabel('DFT Potential')
plt.ylabel('ML Prediction')
plt.title('Optimized Kernel Ridge Regression (Test-set)')
plt.legend()

## Feature Importance

In [ ]:
from sklearn.ensemble import RandomForestRegressor

feat_idx2name={idx:col for idx,col in enumerate(X_train.columns)}
feat_name2idx={col:idx for idx,col in enumerate(X_train.columns)}
feat_name_list=X_train.columns

ntop=20
rf_reg=RandomForestRegressor().fit(X_train_scaled, y_train)
rf_feat_score = rf_reg.feature_importances_
rf_feat_score_sorted_idx=rf_feat_score.argsort()
plt.figure(figsize=(15,12))
rf_top_features=list(map(lambda x:feat_idx2name[x],rf_feat_score_sorted_idx))
plt.barh(rf_top_features[-ntop:],rf_feat_score[rf_feat_score_sorted_idx][-ntop:])
plt.ylabel('Features')
plt.xlabel('Feature Importance Score')
plt.title('Top 20 Features from Random Forest')
plt.show()

## - Model Performance on Top 20 Features (CV)

In [ ]:
from sklearn.kernel_ridge import KernelRidge

print(X_train_scaled.shape,X_test_scaled.shape)
krr_reg=KernelRidge().fit(X_train_scaled, y_train)
y_pred=krr_reg.predict(X_test_scaled)
y_true=y_test
r2=r2_score(y_test,y_pred)
print(f'R^2 = {r2}')
plt.scatter(y_test,y_pred,label=f'R^2 = {r2}')
plt.plot(np.linspace(-2,0),np.linspace(-2,0),ls='dashed',label='X=Y line')
plt.xlabel('DFT Potential')
plt.ylabel('ML Prediction')
plt.title(f'KRR (all features)')
plt.legend()
plt.show()

plt.figure()
ntop=20
rf_top_features_idx=list(map(lambda x:feat_name2idx[x],rf_top_features))
_X_train,_X_test=X_train_scaled.iloc[:,rf_top_features_idx[-ntop:]],X_test_scaled.iloc[:,rf_top_features_idx[-ntop:]]
print(_X_train.shape,_X_test.shape)
krr_reg =KernelRidge().fit(_X_train, y_train)
y_pred=krr_reg.predict(_X_test)
y_true=y_test
r2=r2_score(y_test,y_pred)
print(f'R^2 = {r2}')
plt.scatter(y_test,y_pred,label=f'R^2 = {r2}')
plt.plot(np.linspace(-2,0),np.linspace(-2,0),ls='dashed',label='X=Y line')
plt.xlabel('DFT Potential')
plt.ylabel('ML Prediction')
plt.title(f'KRR (top 20 features)')
plt.legend()


# Multiple Models

## Unoptimized Models

## Optimized Models (Hyper-parameter Tuning)

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import MultiTaskLasso
from sklearn.linear_model import ElasticNet
from sklearn.linear_model import MultiTaskElasticNet
from sklearn.linear_model import Lars
from sklearn.linear_model import LassoLars
from sklearn.linear_model import OrthogonalMatchingPursuit
from sklearn.linear_model import BayesianRidge
from sklearn.linear_model import ARDRegression
from sklearn.linear_model import PassiveAggressiveRegressor
from sklearn.linear_model import RANSACRegressor
from sklearn.linear_model import HuberRegressor
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.kernel_ridge import KernelRidge
from sklearn.svm import SVR
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import ConstantKernel,DotProduct, WhiteKernel,RBF, WhiteKernel, RationalQuadratic, ExpSineSquared
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.experimental import enable_hist_gradient_boosting
from sklearn.ensemble import HistGradientBoostingRegressor
from sklearn.neural_network import MLPRegressor
from sklearn.neighbors import KNeighborsRegressor



MODELS_DICT_UNOPT={'linear_reg':LinearRegression(),
             'ridge':Ridge(),
             'lasso':Lasso(),
#              'multitask_lasso':MultiTaskLasso(),
             'elastic_net':ElasticNet(),
#              'multitask_elastic_net':MultiTaskElasticNet(),
#              'lars':Lars(n_nonzero_coefs=50),
             'lasso_lars':LassoLars(),
             'omp':OrthogonalMatchingPursuit(),
             'bayesian_ridge':BayesianRidge(),
             'ARDR':ARDRegression(),
             'PA':PassiveAggressiveRegressor(),
#              'RANSAC':RANSACRegressor(min_samples=129)),
             'huber':HuberRegressor(epsilon=1,
                                     alpha=1e-3,
                                     tol=1e-6,
                                     max_iter=50000),
#              'polynomial_features':Pipeline([('poly', PolynomialFeatures(degree=3)),
#                                                ('linear', LinearRegression(fit_intercept=True))]),
             'kernel_ridge':KernelRidge(kernel='rbf'),
             'SVR':SVR(),
             'gaussian_process':GaussianProcessRegressor(kernel=DotProduct() + WhiteKernel()),
             'decision_tree':DecisionTreeRegressor(),
             'bagging':BaggingRegressor(),
             'random_forest':RandomForestRegressor(),
             'ada_boost':AdaBoostRegressor(),
             'gradient_boosting_reg':GradientBoostingRegressor(),
#              'hist_gradient_boosting_reg':HistGradientBoostingRegressor(),
             'neural_network':MLPRegressor(hidden_layer_sizes=(50,20),
                                            max_iter=10000),
             'knn_reg':KNeighborsRegressor()
            }

PARAM_GRID_DICT={'linear_reg':{'fit_intercept':[True,False],
                               'normalize':[True,False],
                               'positive':[True,False]
                              },
                 'ridge':{'alpha':[0.5,1,1.5],
                          'tol':[1e-3,1e-4,1e-5],
                          'max_iter':[500,1000,5000],
                          'solver':['auto', 'svd', 'cholesky', 'lsqr', 'sparse_cg', 'sag', 'saga']
                         },
                 'lasso':{'alpha':[1e-2,0.1,0.2,0.5,0.8,1],
                          'positive':[True,False],
                          'tol':[1e-4,1e-5,1e-6,1e-7],
                          'max_iter':[100,200,500,1000],
                         },
    #              'multitask_lasso':MultiTaskLasso(),
                 'elastic_net':{'alpha':[0.5,1,1.5],
                                'l1_ratio':[0.2,0.5,0.8],
                                'positive':[True,False],
                                'tol':[1e-3,1e-4,1e-5],
                                'max_iter':[500,1000,5000]
                               },
    #              'multitask_elastic_net':MultiTaskElasticNet(),
    #              'lars':Lars(n_nonzero_coefs=50),
                 'lasso_lars':{'alpha':[1e-2,0.1,0.2,0.5,0.8,1],
                                'jitter':[None,0.1,1,10,100],
                                'positive':[True,False],
                                'max_iter':[100,200,500,1000],
                               },
                 'omp':{#'n_nonzero_coefs':[None,5,10,20],
    #                   'tol':[0.1,0.01,1e-3,1e-4,1e-5,1e-6,1e-7]
                       },
                 'bayesian_ridge':{'n_iter':[200,300,500],
                                   'tol':[1e-2,1e-3,1e-4],
                                   'alpha_1':[1e-5,1e-6,1e-7],
                                   'alpha_2':[1e-5,1e-6,1e-7],
                                   'lambda_1':[1e-5,1e-6,1e-7],
                                   'lambda_2':[1e-5,1e-6,1e-7],
                                  },
                 'ARDR':{'n_iter':[200,300,500],
                         'tol':[1e-2,1e-3,1e-4],
                         'alpha_1':[1e-5,1e-6,1e-7],
    #                      'alpha_2':[1e-5,1e-6,1e-7],
                         'lambda_1':[1e-5,1e-6,1e-7],
    #                      'lambda_2':[1e-5,1e-6,1e-7],
    #                      'threshold_lambda':[5000,10000,20000]
                        },
                 'PA':{'C':[0.5,0.6,0.7,0.8,0.9,1.0,1.1,1.15],
                       'max_iter':[100,200,500,1000,1500,2000],
                       'tol':[1e-5,1e-6,1e-7,1e-8],
    #                    'early_stopping':[True,False],
    #                    'average':[True,False],
    #                    'validation_fraction':[0.05,0.1,0.2]
                      },
    #              'RANSAC':RANSACRegressor(min_samples=129)),
                 'huber':{'epsilon':[1,1.35,1.5],
                          'alpha':[1e-3,1e-4,1e-5],
                          'tol':[1e-5,1e-6,1e-7],
                          'max_iter':[1000,50000]
                         },
    #              'polynomial_features':Pipeline([('poly', PolynomialFeatures(degree=3)),
    #                                                ('linear', LinearRegression(fit_intercept=True))]),
                 'kernel_ridge':{'kernel': ['chi2', 'linear','polynomial', 'rbf', 'laplacian', 'sigmoid', 'cosine'],
                                 'gamma': [1e-8,1e-7,1e-6,1e-5,1e-3, 1e-4],
                                 'alpha':[0.01,0.1,1,5,10,100]
                                },
                 'SVR':{'kernel':['linear', 'poly', 'rbf', 'sigmoid'],
                        'C':[0.5,1,1.5],
                        'tol':[1e-2,1e-3,1e-4],
                        'gamma':['scale','auto']
                       },
                 'gaussian_process':{'kernel':[DotProduct()+WhiteKernel(),
                                               RBF()+WhiteKernel(),
                                               34.4**2 * RBF(length_scale=41.8)
                                                + 3.27**2 * RBF() * ExpSineSquared()
                                                + 0.446**2 * RationalQuadratic(alpha=17.7, length_scale=0.957)
                                                + 0.197**2 * RBF(length_scale=0.138) + WhiteKernel(noise_level=0.0336),
                                               ConstantKernel(constant_value=1.0, constant_value_bounds=(0.0, 10.0)) * RBF(length_scale=0.5, length_scale_bounds=(0.0, 10.0)) + RBF(length_scale=2.0, length_scale_bounds=(0.0, 10.0))
                                              ],
                                     'alpha':[1e-9,1e-10,1e-11]
                                    },
                 'decision_tree':{#'criterion':['mse', 'friedman_mse', 'mae', 'poisson'],
    #                               'splitter':['best','random'],
    #                               'max_features':['auto', 'sqrt', 'log2'],
    #                               'ccp_alpha':[-1,0,0.1,0.5,1],
                                  'max_depth':[None,10,20,50,100,200]
                                 },
                 'bagging':{'n_estimators': np.arange(50,1000,50),
                            'bootstrap':[True,False],
                            'bootstrap_features':[True,False],
                            'oob_score':[True,False]
                           },
                 'random_forest':{'n_estimators': np.arange(50,501,50),
                                  'bootstrap':[True,False],
                                  'oob_score':[True,False],
    #                               'criterion':['mse', 'mae'],
    #                               'max_features':['auto', 'sqrt', 'log2'],
    #                               'ccp_alpha':[0,0.5,1]
                                  },
                 'ada_boost':{'n_estimators': np.arange(50,501,50),
                              'learning_rate':[0.5,1,1.5],
                              'loss':['linear', 'square', 'exponential']
                             },
                 'gradient_boosting_reg':{#'loss':['ls', 'lad', 'huber', 'quantile'],
    #                                       'learning_rate':[0.5,1,1.5],
                                          'n_estimators': np.arange(50,501,50),
    #                                       'criterion':['friedman_mse', 'mse', 'mae'],
                                          'max_depth':[2,3,10,20,50],
    #                                       'max_features':['auto', 'sqrt', 'log2'],
    #                                       'alpha':[0.6,0.9,1.2],
                                          'tol':[1e-2,1e-3,1e-4],
    #                                       'ccp_alpha':[0.5,1,1.5]
                                         },
    #              'hist_gradient_boosting_reg':{},
                 'neural_network':{'hidden_layer_sizes':[(10,10,10),(50,30,20),(80,60,30)],
                                   'activation':['logistic', 'tanh', 'relu'],
                                   'solver':['lbfgs', 'sgd', 'adam'],
                                   'learning_rate':['constant', 'invscaling', 'adaptive'],
                                   'early_stopping':[True,False]
                                  },
                 'knn_reg':{'n_neighbors':np.arange(1,10,2),
                            'weights':['uniform', 'distance'],
                            'algorithm':['auto', 'ball_tree', 'kd_tree', 'brute'],
                            'leaf_size':[40,50,75]
                           }
                }
print(f'Total Models = {len(PARAM_GRID_DICT)}')
print(f'Total Models = {len(MODELS_DICT_UNOPT)}')

In [112]:
from sklearn.model_selection import cross_validate
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import r2_score,mean_squared_error,mean_absolute_error

def gridSearch(feat_type,X,y,models_dict_unopt,param_grid_dict,read=False):
    outfile_name=f'MODEL_OPT_DF_{feat_type}_{LIBRARY}.pkl'
    if read:
        model_opt_df=getDataFromFile(outfile_name)
        if model_opt_df is not None:
            return model_opt_df
        else:
            pass
        
    gs_results_dict={'feat_set':[],'model_name':[],'opt_model':[],'opt_params':[],}
    scoring='neg_mean_squared_error'
    # test_model='PA'
    # for model_name,param_grid in [(test_model,param_grid_dict[test_model])]:
    for model_name,param_grid in param_grid_dict.items():
        print(model_name)
        unopt_estimator=models_dict_unopt[model_name]
        gs = GridSearchCV(unopt_estimator,
                          param_grid=param_grid,
                          scoring=scoring,
                          refit=True,
                          return_train_score=False,
                          cv=10,
                          verbose=2,
                          n_jobs=-1
                         )
        gs.fit(X, y)
        gs_results_dict['feat_set'].append(f'{feat_type}_{LIBRARY}')
        gs_results_dict['model_name'].append(model_name)
        gs_results_dict['opt_model'].append(gs.best_estimator_)
        gs_results_dict['opt_params'].append(gs.best_params_)
    clear_output()
    model_opt_df=pd.DataFrame.from_dict(gs_results_dict)
    with open(outfile_name,'wb') as outfile:
        pickle.dump(model_opt_df,outfile)
    return model_opt_df

def cv(feat_type,opt_models,X,y,read=False):
    outfile_name=f'CV_RESULT_DF_{feat_type}_{LIBRARY}.pkl'
    if read:
        model_opt_df=getDataFromFile(outfile_name)
        if model_opt_df is not None:
            return model_opt_d
        else:
            pass
    
    num_folds=10
    cv_result_dict={'feat_set':[],'model_name':[],'cv_mean_test_r2':[],'cv_mean_test_mse':[],'cv_mean_test_mae':[]}
    scoring=['r2','neg_mean_squared_error','neg_mean_absolute_error']
    pbar=tqdm(opt_models.iterrows())
    for idx,row in pbar:
        model_name=row['model_name']
        pbar.set_description(model_name)
        model=row['opt_model']
        scores=cross_validate(model,
                              X,
                              y,
                              cv=num_folds,
                              scoring=scoring,
                              n_jobs=-1
                             )
                            
        cv_mean_test_r2=scores['test_r2'].mean()
        cv_mean_test_mse=-scores['test_neg_mean_squared_error'].mean()
        cv_mean_test_mae=-scores['test_neg_mean_absolute_error'].mean()
        cv_result_dict['feat_set'].append(f'{feat_type}_{LIBRARY}')
        cv_result_dict['model_name'].append(model_name)
        cv_result_dict['cv_mean_test_r2'].append(cv_mean_test_r2)
        cv_result_dict['cv_mean_test_mse'].append(cv_mean_test_mse)
        cv_result_dict['cv_mean_test_mae'].append(cv_mean_test_mae)
    cv_result_df=pd.DataFrame.from_dict(cv_result_dict)
    with open(outfile_name,'wb') as outfile:
        pickle.dump(cv_result_df,outfile)
    return cv_result_df

def testErrors(opt_models,X,y,read=False):
    outfile_name=f'TEST_ERROR_DF_{feat_type}_{LIBRARY}.pkl'
    if read:
        model_opt_df=getDataFromFile(outfile_name)
        if model_opt_df is not None:
            return model_opt_df
        else:
            pass
        
    test_error_dict={'feat_set':[],'model_name':[],'test_r2':[],'test_mse':[],'test_mae':[]}
    pbar=tqdm(opt_models.iterrows())
    for idx,row in pbar:
        model_name=row['model_name']
        pbar.set_description(model_name)
        model=row['opt_model']
        pred=model.predict(X)
        test_r2=r2_score(y,pred)
        test_mse=mean_squared_error(y,pred)
        test_mae=mean_absolute_error(y,pred)
        test_error_dict['feat_set'].append(f'{feat_type}_{LIBRARY}')
        test_error_dict['model_name'].append(model_name)
        test_error_dict['test_r2'].append(test_r2)
        test_error_dict['test_mse'].append(test_mse)
        test_error_dict['test_mae'].append(test_mae)
    test_error_df=pd.DataFrame.from_dict(test_error_dict)
    with open(outfile_name,'wb') as outfile:
        pickle.dump(test_error_df,outfile)
    return test_error_df

def getDataFromFile(file_name):
    try:
        with open(file_name, 'rb') as file:
            print('File Found')
            return pickle.load(file)
    except FileNotFoundError:
        print('File Not Found, performe the whole process...')
        return None

def getModelPerformance(feat_type,X_train,y_train,X_test,y_test,read=False):
    outfile_name=f'MODEL_PERFORMANCE_DF_{feat_type}_{LIBRARY}.pkl'
    if read:
        model_performance_df=getDataFromFile(outfile_name)
        if model_performance_df is not None:
            return model_performance_df
        else:
            pass

    _model_opt_df=gridSearch(feat_type,X_train,y_train,
                             MODELS_DICT_UNOPT,PARAM_GRID_DICT,read=read)
    _cv_result_df=cv(feat_type,_model_opt_df,X_train,y_train,read=read)
    _test_error_df=testErrors(_model_opt_df,X_test,y_test,read=read)
    model_performance_df=_model_opt_df.merge(_cv_result_df,how='left',on='model_name')
    model_performance_df=model_performance_df.merge(_test_error_df,how='left',on='model_name')
    with open(outfile_name,'wb') as outfile:
        pickle.dump(model_performance_df,outfile)
    return model_performance_df

In [91]:
%%time

feat_type='2d+3d+fp'

model_performance_df_2d_3d_fp=getModelPerformance(feat_type,
                    X_train_scaled_2d_3d_fp,y_train,
                    X_test_scaled_2d_3d_fp,y_test,
                    read=True                            
                   )

model_performance_df_2d_3d_fp

File Found
CPU times: user 40 ms, sys: 2.22 s, total: 2.26 s
Wall time: 2.26 s


model_name                                          opt_model  \
0              linear_reg    LinearRegression(normalize=True, positive=True)   
1                   ridge        Ridge(alpha=1, max_iter=500, solver='saga')   
2                   lasso                    Lasso(alpha=0.01, max_iter=200)   
3             elastic_net  ElasticNet(alpha=0.5, l1_ratio=0.2, max_iter=500)   
4              lasso_lars                LassoLars(alpha=0.01, max_iter=100)   
5                     omp                        OrthogonalMatchingPursuit()   
6          bayesian_ridge  BayesianRidge(alpha_1=1e-07, alpha_2=1e-05, la...   
7                    ARDR  ARDRegression(alpha_1=1e-05, lambda_1=1e-07, n...   
8                      PA  PassiveAggressiveRegressor(C=0.8, max_iter=100...   
9                   huber  HuberRegressor(alpha=0.001, epsilon=1, max_ite...   
10           kernel_ridge  KernelRidge(alpha=0.01, gamma=0.001, kernel='p...   
11                    SVR            SVR(C=0.5, kernel='linear', tol=0.0001)   
12       gaussian_process  GaussianProcessRegressor(kernel=34.4**2 * RBF(...   
13          decision_tree               DecisionTreeRegressor(max_depth=200)   
14                bagging  (DecisionTreeRegressor(random_state=20995723),...   
15          random_forest  (DecisionTreeRegressor(max_features='auto', ra...   
16              ada_boost  (DecisionTreeRegressor(max_depth=3, random_sta...   
17  gradient_boosting_reg  ([DecisionTreeRegressor(criterion='friedman_ms...   
18         neural_network  MLPRegressor(activation='logistic', hidden_lay...   
19                knn_reg  KNeighborsRegressor(leaf_size=40, weights='dis...   

                                           opt_params  cv_mean_test_r2  \
0   {'fit_intercept': True, 'normalize': True, 'po...       -34.832723   
1   {'alpha': 1, 'max_iter': 500, 'solver': 'saga'...         0.988107   
2   {'alpha': 0.01, 'max_iter': 200, 'positive': F...         0.984050   
3   {'alpha': 0.5, 'l1_ratio': 0.2, 'max_iter': 50...         0.922607   
4   {'alpha': 0.01, 'jitter': None, 'max_iter': 10...         0.934235   
5                                                  {}       -20.540111   
6   {'alpha_1': 1e-07, 'alpha_2': 1e-05, 'lambda_1...         0.987608   
7   {'alpha_1': 1e-05, 'lambda_1': 1e-07, 'n_iter'...         0.978841   
8           {'C': 0.8, 'max_iter': 100, 'tol': 1e-08}         0.915787   
9   {'alpha': 0.001, 'epsilon': 1, 'max_iter': 100...         0.928277   
10  {'alpha': 0.01, 'gamma': 0.001, 'kernel': 'pol...         0.987745   
11  {'C': 0.5, 'gamma': 'scale', 'kernel': 'linear...         0.974348   
12  {'alpha': 1e-10, 'kernel': 34.4**2 * RBF(lengt...         0.987687   
13                                 {'max_depth': 200}         0.932232   
14  {'bootstrap': False, 'bootstrap_features': Tru...         0.962877   
15  {'bootstrap': True, 'n_estimators': 150, 'oob_...         0.961108   
16  {'learning_rate': 0.5, 'loss': 'square', 'n_es...         0.963914   
17  {'max_depth': 3, 'n_estimators': 300, 'tol': 0...         0.973834   
18  {'activation': 'logistic', 'early_stopping': F...         0.963828   
19  {'algorithm': 'auto', 'leaf_size': 40, 'n_neig...         0.935075   

    cv_mean_test_mse  cv_mean_test_mae   test_r2  test_mse  test_mae  
0          14.202987          0.489466  0.838745  0.068351  0.202750  
1           0.004508          0.046330  0.977318  0.009614  0.057180  
2           0.005869          0.059684  0.974728  0.010712  0.071932  
3           0.031899          0.155291  0.927452  0.030751  0.149757  
4           0.026687          0.132405  0.946059  0.022864  0.124048  
5           8.573479          0.918336  0.969307  0.013010  0.087478  
6           0.004669          0.047366  0.972177  0.011793  0.060010  
7           0.007319          0.054729  0.978661  0.009045  0.063581  
8           0.032779          0.115562  0.928988  0.030100  0.105928  
9           0.027042          0.091488  0.937161  0.026636  0.095379 

In [75]:
%%time

feat_type='2d+3d'

model_performance_df_2d_3d=getModelPerformance(feat_type,
                    X_train_scaled_2d_3d,y_train,
                    X_test_scaled_2d_3d,y_test,
                    read=True                            
                   )

model_performance_df_2d_3d

File Found
CPU times: user 49 ms, sys: 9 ms, total: 58 ms
Wall time: 56.1 ms


model_name                                          opt_model  \
0              linear_reg  LinearRegression(fit_intercept=False, normaliz...   
1                   ridge  Ridge(alpha=1.5, max_iter=1000, solver='sag', ...   
2                   lasso                    Lasso(alpha=0.01, max_iter=200)   
3             elastic_net  ElasticNet(alpha=0.5, l1_ratio=0.2, max_iter=500)   
4              lasso_lars                LassoLars(alpha=0.01, max_iter=100)   
5                     omp                        OrthogonalMatchingPursuit()   
6          bayesian_ridge  BayesianRidge(alpha_1=1e-07, alpha_2=1e-05, la...   
7                    ARDR  ARDRegression(alpha_1=1e-07, lambda_1=1e-05, n...   
8                      PA  PassiveAggressiveRegressor(C=1.1, max_iter=200...   
9                   huber  HuberRegressor(alpha=0.001, epsilon=1, max_ite...   
10           kernel_ridge  KernelRidge(alpha=0.01, gamma=0.001, kernel='p...   
11                    SVR              SVR(C=0.5, kernel='linear', tol=0.01)   
12       gaussian_process  GaussianProcessRegressor(alpha=1e-11,\n       ...   
13          decision_tree                            DecisionTreeRegressor()   
14                bagging  (DecisionTreeRegressor(random_state=242741372)...   
15          random_forest  (DecisionTreeRegressor(max_features='auto', ra...   
16              ada_boost  (DecisionTreeRegressor(max_depth=3, random_sta...   
17  gradient_boosting_reg  ([DecisionTreeRegressor(criterion='friedman_ms...   
18         neural_network  MLPRegressor(activation='logistic', early_stop...   
19                knn_reg  KNeighborsRegressor(algorithm='ball_tree', lea...   

                                           opt_params  cv_mean_test_r2  \
0   {'fit_intercept': False, 'normalize': True, 'p...    -18630.653327   
1   {'alpha': 1.5, 'max_iter': 1000, 'solver': 'sa...         0.984071   
2   {'alpha': 0.01, 'max_iter': 200, 'positive': F...         0.984051   
3   {'alpha': 0.5, 'l1_ratio': 0.2, 'max_iter': 50...         0.922607   
4   {'alpha': 0.01, 'jitter': None, 'max_iter': 10...         0.928966   
5                                                  {}         0.924704   
6   {'alpha_1': 1e-07, 'alpha_2': 1e-05, 'lambda_1...         0.984410   
7   {'alpha_1': 1e-07, 'lambda_1': 1e-05, 'n_iter'...         0.976368   
8          {'C': 1.1, 'max_iter': 2000, 'tol': 1e-05}        -1.664085   
9   {'alpha': 0.001, 'epsilon': 1, 'max_iter': 500...        -1.524589   
10  {'alpha': 0.01, 'gamma': 0.001, 'kernel': 'pol...         0.984911   
11  {'C': 0.5, 'gamma': 'scale', 'kernel': 'linear...         0.981242   
12  {'alpha': 1e-11, 'kernel': 34.4**2 * RBF(lengt...         0.986314   
13                                {'max_depth': None}         0.953981   
14  {'bootstrap': False, 'bootstrap_features': Tru...         0.971962   
15  {'bootstrap': False, 'n_estimators': 100, 'oob...         0.970091   
16  {'learning_rate': 0.5, 'loss': 'square', 'n_es...         0.971707   
17  {'max_depth': 3, 'n_estimators': 400, 'tol': 0...         0.977076   
18  {'activation': 'logistic', 'early_stopping': T...         0.977047   
19  {'algorithm': 'ball_tree', 'leaf_size': 40, 'n...         0.956287   

    cv_mean_test_mse  cv_mean_test_mae       test_r2      test_mse  \
0        5322.829784         21.878980 -1.528372e+28  6.478278e+27   
1           0.005865          0.055960  9.858262e-01  6.007839e-03   
2           0.005869          0.059678  9.747346e-01  1.070918e-02   
3           0.031899          0.155291  9.274517e-01  3.075091e-02   
4           0.028048          0.140022  9.435518e-01  2.392657e-02   
5           0.043038          0.075207  9.812164e-01  7.961771e-03   
6           0.005755          0.055186  9.823256e-01  7.491629e-03   
7           0.009102          0.058053  9.856141e-01  6.097719e-03   
8           0.775337          0.326164  9.577918e-01  1.789071e-02   
9           0.723283          0.291469  9.877960e-01  5.172895e-03   
10       

In [ ]:
%%time

feat_type='2d'

model_performance_df_2d=getModelPerformance(feat_type,
                    X_train_scaled_2d,y_train,
                    X_test_scaled_2d,y_test,
                    read=True                            
                   )

model_performance_df_2d

In [113]:
%%time

feat_type='3d'

model_performance_df_3d=getModelPerformance(feat_type,
                    X_train_scaled_3d,y_train,
                    X_test_scaled_3d,y_test,
                    read=False                            
                   )

model_performance_df_3d

knn_reg: : 20it [00:04,  4.44it/s]              
knn_reg: : 20it [00:00, 105.65it/s]              


CPU times: user 1min 18s, sys: 45 s, total: 2min 3s
Wall time: 6min 4s


feat_set_x             model_name  \
0    3d_rdkit             linear_reg   
1    3d_rdkit                  ridge   
2    3d_rdkit                  lasso   
3    3d_rdkit            elastic_net   
4    3d_rdkit             lasso_lars   
5    3d_rdkit                    omp   
6    3d_rdkit         bayesian_ridge   
7    3d_rdkit                   ARDR   
8    3d_rdkit                     PA   
9    3d_rdkit                  huber   
10   3d_rdkit           kernel_ridge   
11   3d_rdkit                    SVR   
12   3d_rdkit       gaussian_process   
13   3d_rdkit          decision_tree   
14   3d_rdkit                bagging   
15   3d_rdkit          random_forest   
16   3d_rdkit              ada_boost   
17   3d_rdkit  gradient_boosting_reg   
18   3d_rdkit         neural_network   
19   3d_rdkit                knn_reg   

                                            opt_model  \
0     LinearRegression(normalize=True, positive=True)   
1   Ridge(alpha=0.5, max_iter=500, solver='sparse_...   
2                        Lasso(alpha=0.01, tol=1e-07)   
3   ElasticNet(alpha=0.5, l1_ratio=0.2, max_iter=5...   
4   LassoLars(alpha=0.01, max_iter=100, positive=T...   
5                         OrthogonalMatchingPursuit()   
6   BayesianRidge(alpha_1=1e-05, alpha_2=1e-07, la...   
7   ARDRegression(alpha_1=1e-07, lambda_1=1e-07, n...   
8   PassiveAggressiveRegressor(C=0.7, max_iter=100...   
9   HuberRegressor(alpha=1e-05, epsilon=1, max_ite...   
10  KernelRidge(alpha=0.01, gamma=0.001, kernel='l...   
11            SVR(C=1.5, kernel='linear', tol=0.0001)   
12  GaussianProcessRegressor(alpha=1e-11,\n       ...   
13                DecisionTreeRegressor(max_depth=20)   
14  (DecisionTreeRegressor(random_state=162716264)...   
15  (DecisionTreeRegressor(max_features='auto', ra...   
16  (DecisionTreeRegressor(max_depth=3, random_sta...   
17  ([DecisionTreeRegressor(criterion='friedman_ms...   
18  MLPRegressor(hidden_layer_sizes=(80, 60, 30), ...   
19  KNeighborsRegressor(leaf_size=40, n_neighbors=...   

                                           opt_params feat_set_y  \
0   {'fit_intercept': True, 'normalize': True, 'po...   3d_rdkit   
1   {'alpha': 0.5, 'max_iter': 500, 'solver': 'spa...   3d_rdkit   
2   {'alpha': 0.01, 'max_iter': 1000, 'positive': ...   3d_rdkit   
3   {'alpha': 0.5, 'l1_ratio': 0.2, 'max_iter': 50...   3d_rdkit   
4   {'alpha': 0.01, 'jitter': None, 'max_iter': 10...   3d_rdkit   
5                                                  {}   3d_rdkit   
6   {'alpha_1': 1e-05, 'alpha_2': 1e-07, 'lambda_1...   3d_rdkit   
7   {'alpha_1': 1e-07, 'lambda_1': 1e-07, 'n_iter'...   3d_rdkit   
8           {'C': 0.7, 'max_iter': 100, 'tol': 1e-08}   3d_rdkit   
9   {'alpha': 1e-05, 'epsilon': 1, 'max_iter': 100...   3d_rdkit   
10  {'alpha': 0.01, 'gamma': 0.001, 'kernel': 'lap...   3d_rdkit   
11  {'C': 1.5, 'gamma': 'scale', 'kernel': 'linear...   3d_rdkit   
12  {'alpha': 1e-11, 'kernel': DotProduct(sigma_0=...   3d_rdkit   
13                                  {'max_depth': 20}   3d_rdkit   
14  {'bootstrap': False, 'bootstrap_features': Tru...   3d_rdkit   
15  {'bootstrap': True, 'n_estimators': 150, 'oob_...   3d_rdkit   
16  {'learning_rate': 1.5, 'loss': 'square', 'n_es...   3d_rdkit   
17  {'max_depth': 3, 'n_estimators': 250, 'tol': 0...   3d_rdkit   
18  {'activation': 'relu', 'early_stopping': False...   3d_rdkit   
19  {'algorithm': 'auto', 'leaf_size': 40, 'n_neig...   3d_rdkit   

    cv_mean_test_r2  cv_mean_test_mse  cv_mean_test_mae  feat_set   test_r2  \
0          0.712830          0.121683          0.276570  3d_rdkit  0.748640   
1          0.721198          0.114029          0.268917  3d_rdkit  0.755759   
2          0.698274          0.127033          0.285227  3d_rdkit  0.733241   
3          0.569582          0.179177          0.362399  3d_rdkit  0.620946   
4          0.612961          0.160159          0.338840  3d_rdkit  0.656023   
5          0.491653          0.211011          0.380081  3d_r

## Best Model Error Vs. Num Features

In [96]:
%%time
from copy import deepcopy
#Perfomance Curve (wrt to no. of features) for best performing model on all features

read=True
outfile_name=f'BEST_MODEL_VS_NUM_FETURES_2d+3d+fp_{LIBRARY}.pkl'
if read:
    model_vs_num_feat_df=getDataFromFile(outfile_name)
    if model_vs_num_feat_df is not None:
        pass
    else:
        best_model=deepcopy(model_performance_df_2d_3d_fp.iloc[\
            model_performance_df_2d_3d_fp['test_mse'].argmin()]['opt_model'])
        best_model_vs_num_feat_dict={'feat_set':[],'model_name':[],'num_feat':[],'test_r2':[],'test_mse':[],'test_mae':[]}
        for num_feat in tqdm(range(1,X_train_scaled_2d_3d_fp.shape[1])):
            _X_train=X_train_scaled_2d_3d_fp.iloc[:,:num_feat]
            _X_test=X_test_scaled_2d_3d_fp.iloc[:,:num_feat]
            best_model=best_model.fit(_X_train,y_train)
            pred=best_model.predict(_X_test)
            test_r2=r2_score(y_test,pred)
            test_mse=mean_squared_error(y_test,pred)
            test_mae=mean_absolute_error(y_test,pred)
            feat_set=set()
            for col in _X_train.columns:
                if '2d' in col:
                    feat_set.add('2d')
                if '3d' in col:
                    feat_set.add('3d')
                if 'fp' in col:
                    feat_set.add('fp')
            feat_set='+'.join(feat_set)+f'_{LIBRARY}'
            best_model_vs_num_feat_dict['feat_set'].append(feat_set)
            best_model_vs_num_feat_dict['model_name'].append(model_name)
            best_model_vs_num_feat_dict['num_feat'].append(num_feat)
            best_model_vs_num_feat_dict['test_r2'].append(test_r2)
            best_model_vs_num_feat_dict['test_mse'].append(test_mse)
            best_model_vs_num_feat_dict['test_mae'].append(test_mae)
        best_model_vs_num_feat_df=pd.DataFrame.from_dict(best_model_vs_num_feat_dict)
        with open(outfile_name,'wb') as outfile:
                pickle.dump(best_model_vs_num_feat_df,outfile)
best_model_vs_num_feat_df

  0%|          | 0/2633 [00:00<?, ?it/s]

File Not Found, performe the whole process...


100%|██████████| 2633/2633 [54:41<00:00,  1.25s/it]

CPU times: user 54min 41s, sys: 337 ms, total: 54min 42s
Wall time: 54min 41s


model_name  num_feat   test_r2  test_mse  test_mae
0    knn_reg         1 -0.621576  0.687334  0.617321
1    knn_reg         2 -0.691283  0.716880  0.616250
2    knn_reg         3  0.520045  0.203437  0.324107
3    knn_reg         4  0.459716  0.229009  0.345179
4    knn_reg         5  0.191475  0.342708  0.405089
5    knn_reg         6  0.230760  0.326056  0.423036
6    knn_reg         7  0.464940  0.226795  0.336250
7    knn_reg         8  0.532399  0.198201  0.301429
8    knn_reg         9  0.459320  0.229177  0.350893

In [106]:
outfile_name=f'BEST_MODEL_VS_NUM_FETURES_2d+3d+fp_{LIBRARY}.pkl'
model_vs_num_feat_df=pd.DataFrame.from_dict(model_vs_num_feat_dict)
with open(outfile_name,'wb') as outfile:
    pickle.dump(model_vs_num_feat_df,outfile)
model_vs_num_feat_df

feat_set model_name  num_feat   test_r2  test_mse  test_mae
0     fp+2d+3d_rdkit    knn_reg         1  0.512375  0.206689  0.260268
1     fp+2d+3d_rdkit    knn_reg         2  0.857394  0.060446  0.160892
2     fp+2d+3d_rdkit    knn_reg         3  0.863277  0.057952  0.160674
3     fp+2d+3d_rdkit    knn_reg         4  0.875495  0.052774  0.164051
4     fp+2d+3d_rdkit    knn_reg         5  0.920836  0.033555  0.116513
...              ...        ...       ...       ...       ...       ...
2628  fp+2d+3d_rdkit    knn_reg      2629  0.984438  0.006596  0.058252
2629  fp+2d+3d_rdkit    knn_reg      2630  0.984590  0.006532  0.057155
2630  fp+2d+3d_rdkit    knn_reg      2631  0.984394  0.006615  0.058140
2631  fp+2d+3d_rdkit    knn_reg      2632  0.983560  0.006969  0.058518
2632  fp+2d+3d_rdkit    knn_reg      2633  0.983358  0.007054  0.059118

[2633 rows x 6 columns]

# Feature Importance

In [ ]:
# feat_idx2name={idx:col for idx,col in enumerate(X_train.columns)}
# feat_name2idx={col:idx for idx,col in enumerate(X_train.columns)}
# feat_name_list=X_train.columns

In [82]:
%%time

#use all models trained on all featurs
#No Need to make correlation plot
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_regression,mutual_info_regression
from sklearn.inspection import permutation_importance

read=True
outfile_name=f'FETURE_IMPORTANCE_DF_2d+3d+fp_{LIBRARY}.pkl'
if read:
    feat_imp_df=getDataFromFile(outfile_name)
    if feat_imp_df is not None:
        pass
    else:
        #Tree-Based
        feat_imp_df=pd.DataFrame()
        feat_cols=X_train_scaled_2d_3d_fp.columns
        feat_imp_df['feat_name']=feat_cols
        for model_name in ['random_forest','ada_boost','gradient_boosting_reg']:
            model=model_performance_df_2d_3d_fp[model_performance_df_2d_3d_fp\
                                                ['model_name']==model_name]['opt_model'].values[0]
            feat_score=model.feature_importances_
            feat_imp_df[f'{model_name}_score']=feat_score


        #K-Best
        kbest_selector=SelectKBest(mutual_info_regression,
                                   k='all').fit(X_train_scaled_2d_3d_fp, y_train)
        feat_score=kbest_selector.scores_
        feat_imp_df['kbest_selector_score']=feat_score


        #Permuation Importance
        pbar=tqdm(model_performance_df_2d_3d_fp.iterrows())
        mean_feat_score=[]
        for idx,row in pbar:
            model_name=row['model_name']
            pbar.set_description(model_name)
            estimator=row['opt_model']
            selector = permutation_importance(estimator,
                                              X_train_scaled_2d_3d_fp, y_train,
                                              scoring='neg_mean_squared_error',
                                              n_jobs=-1)
            feat_score=selector.importances_mean
            mean_feat_score.append(feat_score)
        mean_feat_score=np.array(mean_feat_score).mean(axis=0)
        feat_imp_df['mean_permutation_imp_score']=mean_feat_score
        with open(outfile_name,'wb') as outfile:
                pickle.dump(feat_imp_df,outfile)
feat_imp_df

File Not Found, performe the whole process...


knn_reg: : 20it [40:09, 120.46s/it]              

CPU times: user 22min 55s, sys: 3min 42s, total: 26min 38s
Wall time: 40min 13s


feat_name  random_forest_score  ada_boost_score  \
0        MaxEStateIndex_2d_rdkit         2.426896e-03         0.003081   
1        MinEStateIndex_2d_rdkit         1.809864e-04         0.000000   
2     MaxAbsEStateIndex_2d_rdkit         2.530465e-03         0.004756   
3     MinAbsEStateIndex_2d_rdkit         3.826138e-04         0.001502   
4                   qed_2d_rdkit         4.080384e-05         0.000000   
...                          ...                  ...              ...   
2629               rdkit_fp_2042         8.740769e-07         0.000000   
2630               rdkit_fp_2043         0.000000e+00         0.000000   
2631               rdkit_fp_2044         4.369780e-07         0.000000   
2632               rdkit_fp_2046         1.735681e-05         0.000000   
2633               rdkit_fp_2047         0.000000e+00         0.000000   

      gradient_boosting_reg_score  kbest_selector_score  \
0                        0.002324              1.128163   
1                        0.000017              0.871754   
2                        0.003027              1.128979   
3                        0.000035              0.423935   
4                        0.000001              1.091460   
...                           ...                   ...   
2629                     0.000000              0.243585   
2630                     0.000000              0.000000   
2631                     0.000000              0.295205   
2632                     0.000000              0.360974   
2633                     0.000000              0.029960   

      mean_permutation_imp_score  
0                   1.229062e-03  
1                   1.552445e-04  
2                   1.043425e-03  
3                   5.609654e-04  
4                   3.379208e-04  
...                          ...  
2629                3.400861e-05  
2630               -3.658335e-19  
2631                3.076667e-05  
2632                3.999575e-05  
2633                5.035610e-06  

[2634 rows x 6 columns]

In [ ]:
# %%time
# from sklearn.feature_selection import RFECV

# pbar=tqdm(model_opt_data_df.iterrows())
# for idx,row in pbar:
#     model_name=row['model_name']
#     pbar.set_description(model_name)
#     if model_name in ['huber','kernel_ridge','gaussian_process','bagging','neural_network','knn_reg']:
#         continue
#     estimator=row['best_model']
#     selector = RFECV(estimator,
#                      scoring='neg_mean_squared_error',
#                      cv=10,
#                      n_jobs=-1)
#     selector = selector.fit(X_train_scaled, y_train)
#     feat_score=selector.ranking_
#     feat_score_sorted_idx=feat_score.argsort()
#     feat_imp_dict[f'{model_name}_rfecv']=feat_score_sorted_idx



# feat_imp_df=pd.DataFrame.from_dict(feat_imp_dict)
# feat_imp_df
# from scipy import stats

# rfecv_cols=[col for col in feat_imp_df.columns if 'rfecv' in col]

# feat_ranking=[]
# for idx,row in feat_imp_df[rfecv_cols].iterrows():
#     mode=row.mode().values
#     feat=None
#     if len(mode)==1:
#         feat=mode[0]
#     else:
#         feat=row['random_forest_rfecv']
#     feat_ranking.append(feat)
# features=list(map(lambda x:feat_idx2name[x],feat_ranking))
# features[:20]

# perm_imp_cols=[col for col in feat_imp_df.columns if 'perm_imp' in col]

# feat_ranking=[]
# for idx,row in feat_imp_df[perm_imp_cols].iterrows():
#     mode=row.mode().values
#     feat=None
#     if len(mode)==1:
#         feat=mode[0]
#     else:
#         feat=row['random_forest_perm_imp']
#     feat_ranking.append(feat)
# features=list(map(lambda x:feat_idx2name[x],feat_ranking))
# features[:22]

# Top 20 features (re-train optimized models)

In [97]:
%%time

#choose model which perform best on all features
read=True
outfile_name=f'TOP_{ntop}_FEAT_MODEL_PERFORMANCE_DF_{LIBRARY}.pkl'
if read:
    top_feat_model_performance_df=getDataFromFile(outfile_name)
    if top_feat_model_performance_df is not None:
        pass
    else:
        ntop=20
        top_feat_cols=feat_imp_df.sort_values(by='random_forest_score')['feat_name'][:ntop]
        X_train_top_feat=X_train_scaled_2d_3d_fp[feat_cols]
        X_test_top_feat=X_test_scaled_2d_3d_fp[feat_cols]
        feat_set=f'top_{ntop}_feat'
        top_feat_model_performance_df=getModelPerformance(feat_set,
                            X_train_top_feat,y_train,
                            X_test_top_feat,y_test,
                            read=False
                           )
        with open(outfile_name,'wb') as outfile:
                pickle.dump(top_feat_model_performance_df,outfile)
        print(X_train_top_feat.shape,X_test_top_feat.shape)
top_feat_model_performance_df

ValueError: arrays must all be same length

# Best Model Error per Class of Compounds

In [121]:
%%time
#***ONLY BEST MODEL***

best_model_row=model_performance_df_2d_3d_fp.iloc[model_performance_df_2d_3d_fp['test_mse'].argmin()]
best_model=best_model_row['opt_model']
best_model_name=best_model_row['model_name']

pred_train=best_model.predict(X_train_scaled_2d_3d_fp)
best_model_error_per_comp_train_df=pd.DataFrame()
best_model_error_per_comp_train_df['FG_name']=data_train['FG_name']
best_model_error_per_comp_train_df['true_y']=data_train['Redox Potential']
best_model_error_per_comp_train_df['pred_y']=pred_train
best_model_error_per_comp_train_df['abs_error']=np.abs(best_model_error_per_comp_train_df['true_y']-best_model_error_per_comp_train_df['pred_y'])
best_model_error_per_comp_train_df['%abs_error']=best_model_error_per_comp_train_df['abs_error']*100/np.abs(best_model_error_per_comp_train_df['true_y'])
gb_train=best_model_error_per_comp_train_df.groupby('FG_name').mean().sort_values(by='abs_error')
gb_train['model_name']=best_model_name
gb_train['feat_set']='2d+3d+fp_'+LIBRARY
with open(f'BEST_MODEL_MEAN_ERROR_PER_COMPOUND_TRAIN_{LIBRARY}.pkl','wb') as outfile:
    pickle.dump(gb_train,outfile)

pred_test=best_model.predict(X_test_scaled_2d_3d_fp)
best_model_error_per_comp_test_df=pd.DataFrame()
best_model_error_per_comp_test_df['FG_name']=data_test['FG_name']
best_model_error_per_comp_test_df['true_y']=data_test['Redox Potential']
best_model_error_per_comp_test_df['pred_y']=pred_test
best_model_error_per_comp_test_df['abs_error']=np.abs(best_model_error_per_comp_test_df['true_y']-best_model_error_per_comp_test_df['pred_y'])
best_model_error_per_comp_test_df['%abs_error']=best_model_error_per_comp_test_df['abs_error']*100/np.abs(best_model_error_per_comp_test_df['true_y'])
gb_test=best_model_error_per_comp_test_df.groupby('FG_name').mean().sort_values(by='abs_error')
gb_test['model_name']=best_model_name
gb_test['feat_set']='2d+3d+fp_'+LIBRARY
with open(f'BEST_MODEL_MEAN_ERROR_PER_COMPOUND_TEST_{LIBRARY}.pkl','wb') as outfile:
    pickle.dump(gb_test,outfile)
    
gb_test,gb_test

CPU times: user 78 ms, sys: 0 ns, total: 78 ms
Wall time: 77.3 ms


(           true_y    pred_y  abs_error  %abs_error     model_name  \
 FG_name                                                             
 F       -1.630000 -1.631667   0.001667    0.102249  random_forest   
 OCH3    -1.750000 -1.754067   0.004067    0.232381  random_forest   
 SH      -1.720000 -1.726533   0.006533    0.379845  random_forest   
 P(CH3)2 -1.750000 -1.765867   0.015867    0.906667  random_forest   
 S(CH3)2 -1.770000 -1.752267   0.017733    1.001883  random_forest   
 CH2=CH  -1.730000 -1.689933   0.040067    2.315992  random_forest   
 C6H5    -1.745000 -1.704667   0.040333    2.306479  random_forest   
 CN      -0.748462 -0.779299   0.047749   49.298974  random_forest   
 NH2     -2.092941 -2.091404   0.050808    2.375292  random_forest   
 Cl      -1.610000 -1.669933   0.059933    3.722567  random_forest   
 COCH3   -1.580000 -1.513600   0.066400    4.202532  random_forest   
 NO2     -0.560000 -0.510014   0.077653   18.847869  random_forest   
 OH      -1.835000 -

# Features


* **SlogP_VSA1 - SlogP_VSA12** - Sum of van der Waals surface area (in Å2) over all atoms such that SlogP is in a specified range [[1](http://www.cadaster.eu/sites/cadaster.eu/files/challenge/descr.htm)]


* **SlogP** - Log of the octanol/water partition coefficient (including implicit hydrogens). This property is an atomic contribution model [[Crippen 1999](https://pubs.acs.org/doi/full/10.1021/ci990307l)] that calculates logP from the given structure; i.e., the correct protonation state (washed structures). Results may vary from the logP(o/w) descriptor. The training set for SlogP was ~7000 structures.


* **fr_NH0** - Number of Tertiary amines


* **fr_NH2** - Number of Primary amines


* **fr_ArN** - Number of N functional groups attached to aromatics


* **fr_aniline** - Number of anilines


* **VSA_EState1 - VSA_EState10** MOE-type descriptors using EState indices and surface area contributions (developed at RD, not described in the CCG paper)[[2](http://www.rdkit.org/docs/GettingStartedInPython.html#list-of-available-descriptors)]


* **E-statesindex** - Electrotopological State index for atoms in a molecule. Encode the topology and electronic environment of molecular fragments. This index combines both the electronic character and the topological environment of each skeletal atom in a molecule [[3](https://pubs.acs.org/doi/pdf/10.1021/ci00001a012),[4](https://www.researchgate.net/publication/224004106_Electrotopological_State_Atom_E-State_Index_in_Drug_Design_QSAR_Property_Prediction_and_Toxicity_Assessment),[5](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6147309/)]


* **HallKierAlpha** - Ratio of non-C(sp3) atom covalent radii to the covalent radius of sp3 carbon summed over all atoms in a molecule [[6](https://onlinelibrary.wiley.com/doi/abs/10.1002/9780470125793.ch9)]


* **Chi(1/2)n** - Atomic connectivity index (order 1/2) in the hydrogen suppressed graph .This is calculated as the sum of 1/sqrt(didj) over all bonds between heavy atoms i and j where i < j. di (number of heavy neighbors)


* **Ipc** - information content of the coefficients of the characteristic polynomial of the adjacency matrix of a hydrogen-suppressed graph of a molecule


* **Chi3v** - Atomic valence connectivity index (order 3). This is calculated as the sum of 1/sqrt(vivj) over all bonds between heavy atoms i and j where i < j. For a heavy atom i let vi = (pi - hi) / (Zi - pi - 1) where pi is the number of s and p valence electrons of atom i.


* **BertzCT** - A topological index meant to quantify "complexity" of molecules. Consists of a sum of two terms, one representing the complexity of the bonding, the other representing the complexity of the distribution of heteroatoms (from graph theory)


* **Kappa3** - Third kappa shape index: (n-1) (n-3)2 / p32 for odd n, and (n-3) (n-2)2 / p32 for even n. n denotes the number of atoms in the hydrogen suppressed graph. Also, let p2 denote the number of paths of length 2 and p3 the number of paths of length 3


* **PEOE_VSA9** - calculated using partial using partial charges and van der waals surface area contributions of atoms


* **labuteASA** - Water accessible surface area calculated using a radius of 1.4 A for the water molecule. A polyhedral representation is used for each atom in calculating the surface area.


* **SMR_VSA(0-7)** - based on an approximate accessible van der Waals surface area (in Å2) and contribution to Molar Refractivity, summed over all atoms


* **MinAbsPartialCharge** - self-explanatory


* **MaxAbsEStateIndex** - Maximum Absolute Electrotopological State index of molecule


* **balabanJ** - Balaban's connectivity topological index

### References: 
1. http://www.cadaster.eu/sites/cadaster.eu/files/challenge/descr.htm, https://www.juit.ac.in/attachments/biomin/help.html


2. http://www.rdkit.org/docs/GettingStartedInPython.html#list-of-available-descriptors


3. https://pubs.acs.org/doi/pdf/10.1021/ci00001a012


4. https://www.researchgate.net/publication/224004106_Electrotopological_State_Atom_E-State_Index_in_Drug_Design_QSAR_Property_Prediction_and_Toxicity_Assessment


5. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6147309/


6. Hall, L.H., Kier, L.B.; The Molecular Connectivity Chi Indices and Kappa Shape Indices in Structure-Property Modeling; Reviews of Computational Chemistry 2


* [Crippen 1999] - Wildman, S.A., Crippen, G.M.; Prediction of Physiochemical Parameters by Atomic Contributions; J. Chem. Inf. Comput. Sci. 39 No. 5 (1999) 868–873




* **Graph Descriptors** - https://github.com/rdkit/rdkit-orig/blob/master/rdkit/Chem/GraphDescriptors.py


* **MOE** - MOE (Molecular Operating Environment) 2011 software were used to calculate about 300 molecular descriptors for each of these compounds


* **SMR** - molar refractivity (MR), a common descriptor accounting for molecular size and polarizability (https://pubs.acs.org/doi/full/10.1021/ci990307l)

* https://www.researchgate.net/figure/Molecular-descriptors-used-and-their-explanation-adapted-from-MOE-tutorial-Chemical_tbl1_225274017

* **fr** - http://rdkit.org/docs/source/rdkit.Chem.Fragments.html

* **BertzCT** - https://pubs.acs.org/doi/pdf/10.1021/ja00402a071

* **balabanJ** - https://www.sciencedirect.com/science/article/pii/0009261482800092
